File: backend/app/services/volume.py
"""Authoritative volume / confidence / status computation.

Volume is sourced from the weighbridge (tarozi) only: weight ÷ density.
Camera-based dimension measurement is not part of this calculation — the
camera is only used for plate/material recognition (see `services.detection`).

1:1 port of the TS preview in `packages/calc`. This is the single source of
truth — the API recomputes on every save and never trusts client-supplied
volume/status. Parity with the TS port is enforced by `tests/test_volume.py`
(shared fixtures with packages/calc/src/volume.test.ts).
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class MaterialSpec:
    lo: float  # density range low
    hi: float  # density range high


@dataclass(frozen=True)
class VolumeInput:
    density: float  # rho, t/m³
    weight_kg: float  # wkg


@dataclass(frozen=True)
class VolumeResult:
    volume_final: float  # Vf / m3
    confidence: float  # vConf %
    status: str  # confirm | flagged | inspect


def _round2(n: float) -> float:
    return round(n * 100) / 100


def compute_volume(inp: VolumeInput, material: MaterialSpec) -> VolumeResult:
    rho = inp.density or 0.0
    wt = (inp.weight_kg or 0.0) / 1000.0  # tonnes

    vf = wt / rho if rho > 0 else 0.0
    rho_ok = material.lo - 0.06 <= rho <= material.hi + 0.06

    if rho <= 0 or wt <= 0:
        status = "inspect"
        confidence = 0.0
    elif not rho_ok:
        status = "flagged"
        confidence = 80.0
    else:
        status = "confirm"
        confidence = 96.0

    return VolumeResult(
        volume_final=_round2(vf),
        confidence=confidence,
        status=status,
    )

File: packages/calc/src/volume.ts
// Volume / confidence / status computation — TS port of the backend formula.
//
// Volume is sourced from the weighbridge (tarozi) only: weight ÷ density.
// Camera-based dimension measurement is not part of this calculation.
//
// IMPORTANT: this is a PREVIEW port for the operator UI only. The backend
// (`services/volume.py`) is authoritative and recomputes on save. A parity test
// keeps the two in sync (see volume.test.ts and backend tests/test_volume.py).

import type { StatusKey } from '@karier/types';

import { materialById } from './materials';

export interface VolumeInput {
  materialId: string;
  density: number; // rho, t/m³
  weightKg: number; // wkg
}

export interface VolumeResult {
  volumeFinal: number; // Vf / m3
  confidence: number; // vConf %
  status: StatusKey;
}

const round2 = (n: number): number => Math.round(n * 100) / 100;

export function computeVolume(input: VolumeInput): VolumeResult {
  const mat = materialById(input.materialId);
  const rho = input.density || 0;
  const wt = (input.weightKg || 0) / 1000; // tonnes

  const vf = rho > 0 ? wt / rho : 0;

  const lo = mat?.lo ?? 0;
  const hi = mat?.hi ?? 9;
  const rhoOk = rho >= lo - 0.06 && rho <= hi + 0.06;

  let status: StatusKey;
  let confidence: number;
  if (rho <= 0 || wt <= 0) {
    status = 'inspect';
    confidence = 0;
  } else if (!rhoOk) {
    status = 'flagged';
    confidence = 80;
  } else {
    status = 'confirm';
    confidence = 96;
  }

  return { volumeFinal: round2(vf), confidence, status };
}
